lib/src/default_revset_engine.rs

```diff
@@ -825,9 +825,14 @@ impl<'index, 'heads> EvaluationContext<'index, 'heads> {
         S: InternalRevset<'a> + ?Sized,
         T: InternalRevset<'b> + ?Sized,
     {
-        let mut reachable: HashSet<_> = root_set.iter().map(|entry| entry.position()).collect();
+        let root_positions = root_set.iter().map(|entry| entry.position()).collect_vec();
+        let bottom_position = *root_positions.last().unwrap_or(&IndexPosition::MAX);
+        let mut reachable: HashSet<_> = root_positions.into_iter().collect();
         let mut index_entries = vec![];
-        let candidates = self.walk_ancestors(head_set).collect_vec();
+        let candidates = self
+            .walk_ancestors(head_set)
+            .take_while(|entry| entry.position() >= bottom_position)
```

Comment:
I haven't considered that, but I think it would require a bit more walk if the condition were `.take_while(|entry| entry.generation_number() >= min_generation)`. `entry.next()` could have the same generation number even if it were `min(roots)`.

Reply:
Let's say the graph looks like this (sorted by index position):

```
F pos=6, gen=4
|
E pos=5, gen=3
|
D pos=4, gen=2
|
| C pos=3, gen=3
| |
| B pos=2, gen=2
| /
A pos=1, gen=1
```

If we now asked for B:C, we would stop after B either way. Same with E:F (stop after E). For C:F, we could skip D only if we use generation numbers. For D:C, we can skip B only if we use position numbers. Note that for B:F, which is kind of the reverse of D:C, we can't skip any commits when using position number (or generation numbers). This highlights what I don't like about position numbers: they're not a property of the graph but a property of the index, which means the runtime depends on which order we indexed commits. I don't know which choice is generally better in practice.

Reply:
> For C:F, we could skip D only if we use generation numbers.

Since we still need to visit `C`, we can't break at `D`. I think, in order to eliminate uninteresting nodes based on the graph topology, we should, for each linear branch, avoid adding an entry to the "to visit" queue if it's included in the roots. Since `(D|E):` should include `E`, we would need to calculate `roots(D|E)` to build unwanted entries (inclusive.)

> This highlights what I don't like about position numbers: they're not a property of the graph but a property of the index, which means the runtime depends on which order we indexed commits.

True, but here we're constructing the `descendants(roots)` graph property, and we only know the `ancestors(whatever)` topology. The position (or revision number in Mercurial) is merely a hint for early termination.

Reply:
> Since we still need to visit `C`, we can't break at `D`.

We calculate `C:F` by walking from F, right? So we'd find F, then E, then D, and since D has lower generation number than C, we can stop walking.

Anyway, this PR is already an improvement, and I'm not sure whether generation numbers perform better in practice, so I won't block this PR on that.

Reply:
> We calculate `C:F` by walking from F, right? So we'd find F, then E, then D, and since D has lower generation number than C, we can stop walking.

Ah, yes. I was blind. However, with the following graph, we'll need to visit `C` for `C:(F|G)`.
```
  G pos=7, gen=4 (parents = [C])
  |
F pos=6, gen=4
| |
E pos=5, gen=3
| |
D pos=4, gen=2
| |
| C pos=3, gen=3
| |
| B pos=2, gen=2
| /
A pos=1, gen=1
```

I think `min(generation_numbers)` condition could be applied to the `RevWalk`'s queue, not to the iterator. `if entry.generation_number() >= min_generation_number { push_wanted(..) }`. I'll add a comment about that.